lib/assert.js

```diff
@@ -608,6 +612,139 @@ assert.notStrictEqual = function notStrictEqual(actual, expected, message) {
   }
 };
 
+/**
+ * Compares two objects or values recursively to check if they are equal.
+ * @param {any} actual - The actual value to compare.
+ * @param {any} expected - The expected value to compare.
+ * @param {boolean} [loose=false] - Whether to use loose comparison (==) or strict comparison (===). Defaults to false.
+ * @param {Set} [comparedObjects=new Set()] - Set to track compared objects for handling circular references.
+ * @returns {boolean} - Returns `true` if the actual value matches the expected value, otherwise `false`.
+ * @example
+ * // Loose comparison (default)
+ * compareBranch({a: 1, b: 2}, {a: 1, b: '2'}); // true
+ *
+ * // Strict comparison
+ * compareBranch({a: 1, b: 2}, {a: 1, b: 2}, true); // true
+ */
+function compareBranch(
+  actual,
+  expected,
+  loose = false,
+  comparedObjects = new Set(),
+) {
+  function isPlainObject(obj) {
+    return typeof obj === 'object' && obj.constructor === Object;
+  }
```

Comment:
It's going to throw if `obj` is `null`, and `.constructor` could be a own property
```suggestion
  function isPlainObject(obj) {
    if (typeof obj !== 'object' || obj === null) return false;
    const proto = ObjectGetPrototypeOf(obj);
    return proto === Object || proto === null;
  }
```